crates/sui-core/src/node_sync/node_follower.rs

```diff
@@ -2,797 +2,538 @@
-        self.effects_stake_map.remove(digests);
-        self.effects_vote_map.remove(digests);
-    }
-}
+async fn follower_process<A, Handler>(
+    node_sync_handle: Handler,
+    node_sync_store: Arc<NodeSyncStore>,
+    aggregator: Arc<AuthorityAggregator<A>>,
+    max_stream_items: u64,
+    metrics: &GossipMetrics,
+    mut cancel_receiver: oneshot::Receiver<()>,
+) where
+    A: AuthorityAPI + Send + Sync + 'static + Clone,
+    Handler: SyncHandler + Clone,
+{
+    debug!("starting follower process");
 
-/// Waiter is used to single-shot concurrent requests and wait for dependencies to finish.
-struct Waiter<Key, ResultT> {
-    waiters: Mutex<HashMap<Key, broadcast::Sender<ResultT>>>,
-}
+    let mut backoff = HashMap::new();
 
-impl<Key, ResultT> Waiter<Key, ResultT>
-where
-    Key: std::hash::Hash + Eq + Clone + std::fmt::Debug,
-    ResultT: Clone,
-{
-    fn new() -> Self {
-        Self {
-            waiters: Mutex::new(HashMap::new()),
-        }
-    }
+    loop {
+        let mut follower_tasks = FuturesUnordered::new();
+        let mut reconnects = FuturesUnordered::new();
 
-    /// Returns (true, rx) if there are no other waiters yet, or else (false, rx).
-    /// All rxes can be woken by calling notify(key, result)
-    async fn wait(&self, key: &Key) -> (bool, broadcast::Receiver<ResultT>) {
-        let waiters = &mut self.waiters.lock().unwrap();
-        let entry = waiters.entry(key.clone());
-
-        match entry {
-            hash_map::Entry::Occupied(e) => (false, e.get().subscribe()),
-            hash_map::Entry::Vacant(e) => {
-                let (tx, rx) = broadcast::channel(1);
-                e.insert(tx);
-                (true, rx)
+        let start_one_task = |name, handle, store| {
+            let start_time = Instant::now();
+            let client = aggregator.clone_client(name);
+            async move {
+                let result =
+                    follow_one_peer(handle, store, *name, client, max_stream_items, metrics)
+                        .await
+                        .tap_err(|e| warn!("follower task exited with error {}", e));
+                (result, start_time, name)
             }
-        }
-    }
+        };
 
-    async fn notify(&self, key: &Key, res: ResultT) -> SuiResult {
-        if let Some(tx) = self.waiters.lock().unwrap().remove(key) {
-            tx.send(res).map_err(|_| SuiError::GenericAuthorityError {
-                error: format!("couldn't notify waiters for key {:?}", key),
-            })?;
-        } else {
-            trace!("no pending waiters");
+        for (name, _) in aggregator.committee.members() {
+            follower_tasks.push(start_one_task(
+                name,
+                node_sync_handle.clone(),
+                node_sync_store.clone(),
+            ));
         }
-        Ok(())
-    }
-}
 
-struct DigestsMessage {
-    sync_arg: SyncArg,
-    tx: Option<oneshot::Sender<SuiResult>>,
-}
+        loop {
```

Comment:
I'm confused - this code does have sleeps and delays. (Search for `MIN_RECONNECTION_DELAY`). Can you explain the scenario you are worried about?

Reply:
the `MIN_RECONNECTION_DELAY ` is not sufficient, We need a sleep statement in both the outer loop and the inner loop that is guaranteed to be hit at least every once in a while, because these are both infinite loops. Without sleep statements, once a tread starts execution of the loop, it will never be released back to the runtime.